### PR TITLE
feat(web): enables mouse interactivity for the predictive banner 🧵

### PR DESCRIPTION
**Pull request by jahorton**
Follows from #5665.

Addresses part of #3963, integrating 15.0's abstraction for OSK event handling into the 12.0 `SuggestionBanner` and `UITouchHandlerBase`.  This may only be one aspect of the issue, but it's a major step in the right direction.  Also includes some (required-for-this) polish that I knew needed to be done anyway for `InputEventEngine`, as it was a bit too specific.

As hinted by the title, the predictive-text banner now accepts mouse-based interaction when on devices without touch enabled.  This will be of particular use for #5723, but it's also a notable step toward supporting predictive text more broadly and openly within KMW.

### SUITE_REGRESSION - ensure that standard predictive-text banner operations continue as normal

**GROUP_IPHONESE** - test on a Simulator-based iPhone SE, within the Keyman app
**GROUP_GALAXY_FOLD** - test on a Chrome-emulated "Galaxy Fold" (if not available, "Responsive" 280 x 653) on the "Prediction - robust testing" test page

(These are selected because they're among the least-wide [narrowest] test target setups that I know.)

**TEST_LONG_SUGGESTION** - ensure long suggestions are scrollable and may be applied
- type `internat` - `internationally` should appear as the second suggestion.
- As that's a fairly long word, you should be able to horizontally scroll it.  Make sure that's possible and that it works intuitively.
- Ensure you can also tap it to apply the suggestion

### SUITE_MOUSE_USE - tests new functionality

**GROUP_DESKTOP** - use the "Prediction - robust testing" test page _without emulating a mobile device_.  When the OSK is active, you should see something like this:

image

Yes, it's **ugly**.  We know, but we're not here to test that.  You may also want to resize the OSK to a similar size to facilitate the following test:

**TEST_LONG_SUGGESTION** - ensure long suggestions are scrollable and may be applied.
- type `internat` - `internationally` should appear as the second suggestion.
- As that's a fairly long word, you should be able to horizontally scroll it.  
    - If the full word is visible, your OSK is too wide - resize the OSK to something narrower.
    - Make sure it may be scrolled horizontally and that this works intuitively.
- Shows up easily on iPhone SE
- Ensure you can also tap `internationally` to apply the suggestion.

### CODE REVIEW

File: web/source/osk/uiTouchHandlerBase.ts [hunk 1]
```diff
@@ -425,15 +409,16 @@ namespace com.keyman.osk {
...
-        let parent = base.parentElement; // parentElement is _Box
-        let top = base.offsetTop;
-        if(parent) {
-          top += parent.offsetTop;
-        }
+        //let parent = base.parentElement; // parentElement is _Box
+        // let top = base.offsetTop;
+        // if(parent) {
+        //   top += parent.offsetTop;
+        // }
```
Comment (mcdurdin) [inline web/source/osk/uiTouchHandlerBase.ts:416]: Please remove commented code?

**Review by mcdurdin — APPROVED** [review]
Changes LGTM.

File: web/source/osk/mouseEventEngine.ts [hunk 2]
```diff
@@ -9,27 +9,56 @@ namespace com.keyman.osk {
...
-    public constructor(vkbd: VisualKeyboard) {
-      // document.body is the event root b/c we need to track the mouse if it leaves
-      // the VisualKeyboard's hierarchy.
-      super(vkbd, document.body);
+    public constructor(config: InputEventEngineConfig) {
+      super(config);
 
       this._mouseStart = this.onMouseStart.bind(this);
       this._mouseMove  = this.onMouseMove.bind(this);
       this._mouseEnd   = this.onMouseEnd.bind(this);
```
Comment (darcywong00) [inline web/source/osk/mouseEventEngine.ts:17]: I take it mouseEventEngine.ts currently only handles left mouse clicks. Could a future feature allow right-click (on a suggestion) for a "copy" option? Just wondering if later on we would add right and middle clicks...

Comment (jahorton) [reply web/source/osk/mouseEventEngine.ts:17]: Maybe?  It'd definitely be a "future feature", though.

We'd probably want to include something similar for touch - I know that simultaneous double-touch and triple-touch act as analogies for the other mouse clicks on some devices, so it wouldn't be unreasonable to extend the abstraction to incorporate that if desired.

**Comment by MakaraSok** [comment]
## SUITE_REGRESSION: - ensure that standard predictive-text banner operations continue as normal
* GROUP_IPHONESE: - test on a Simulator-based iPhone SE, within the Keyman app
  - **TEST_LONG_SUGGESTION: PASSED** The long suggestion can be long-tapped on and swipe left or right to scroll and see the rest of the word. It seems intuitive, but one might not know the trick. 

* GROUP_GALAXY_FOLD: - test on a Chrome-emulated "Galaxy Fold" (if not available, "Responsive" 280 x 653) on the "Prediction - robust testing" test page
  - **TEST_LONG_SUGGESTION: PASSED** Similar to the above test case -- the long word can be long-tapped on to swipe left or right to see the rest of the word. 


## SUITE_MOUSE_USE: - tests new functionality
* GROUP_DESKTOP: - use the "Prediction - robust testing" test page _without emulating a mobile device_.  When the OSK is active, you should see something like this:
  - **TEST_LONG_SUGGESTION: PASSED** The suggestion does show up and it is possible to click and drag on the long suggestion to see the rest of the word.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.118-alpha](https://keyman.com/downloads/releases/alpha/15.0.118)

**Pull request merged**